js/buffer.ts

```diff
@@ -0,0 +1,219 @@
+// This code has been ported almost directly from Go's src/bytes/buffer.go
+// Copyright 2009 The Go Authors. All rights reserved. BSD license.
+// https://github.com/golang/go/blob/master/LICENSE
+
+//import * as io from "./io";
+import { Reader, Writer, ReadResult } from "./io";
+import { assert } from "./util";
+import { TypedArray } from "./types";
+import { TextDecoder } from "./text_encoding";
+
+// MIN_READ is the minimum ArrayBuffer size passed to a read call by
+// buffer.ReadFrom. As long as the Buffer has at least MIN_READ bytes beyond
+// what is required to hold the contents of r, readFrom() will not grow the
+// underlying buffer.
+const MIN_READ = 512;
+
+// `off` is the offset into `dst` where it will at which to begin writing values
+// from `src`.
+// Returns the number of bytes copied.
+function copyBytes(dst: TypedArray, src: TypedArray, off = 0): number {
+  const r = dst.byteLength - off;
+  if (src.byteLength > r) {
+    src = src.subarray(0, r);
+  }
+  dst.set(src, off);
+  return src.byteLength;
+}
+
+/** A Buffer is a variable-sized buffer of bytes with read() and write()
+ * methods. Based on https://golang.org/pkg/bytes/#Buffer
+ */
+export class Buffer implements Reader, Writer {
+  private buf: Uint8Array; // contents are the bytes buf[off : len(buf)]
+  private off = 0; // read at buf[off], write at buf[buf.byteLength]
+
+  constructor(ab?: ArrayBuffer) {
+    if (ab == null) {
+      this.buf = new Uint8Array(0);
+    } else {
+      this.buf = new Uint8Array(ab, 0, ab.byteLength);
+    }
+  }
+
+  /** bytes() returns a slice holding the unread portion of the buffer.
+   * The slice is valid for use only until the next buffer modification (that
+   * is, only until the next call to a method like read(), write(), reset(), or
+   * truncate()). The slice aliases the buffer content at least until the next
+   * buffer modification, so immediate changes to the slice will affect the
+   * result of future reads.
+   */
+  bytes(): Uint8Array {
+    return this.buf.subarray(this.off);
+  }
+
+  /** toString() returns the contents of the unread portion of the buffer
+   * as a string.
+   */
+  toString(): string {
+    const decoder = new TextDecoder();
+    return decoder.decode(this.buf.subarray(this.off));
+  }
+
+  /** empty() returns whether the unread portion of the buffer is empty. */
+  empty() {
+    return this.buf.byteLength <= this.off;
+  }
+
+  /** length is a getter that returns the number of bytes of the unread
+   * portion of the buffer
+   */
+  get length() {
+    return this.buf.byteLength - this.off;
+  }
+
+  /** Cap returns the capacity of the buffer's underlying byte slice, that is,
+   * the total space allocated for the buffer's data.
+   */
+  cap(): number {
```

Comment:
1. why is `length` a getter but `cap` is a function?
2. i think `capacity` is more explicit.